server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -0,0 +1,403 @@
+     */
+    record RerouteInput(RoutingAllocation routingAllocation, List<ShardRouting> ignoredShards) {}
+
+    private void updateDesiredBalanceAndReroute(RerouteInput rerouteInput) {
+
+        final var routingAllocation = rerouteInput.routingAllocation().mutableCloneForSimulation();
+        final var routingNodes = routingAllocation.routingNodes();
+        final var ignoredShards = new HashSet<>(rerouteInput.ignoredShards());
+        final var desiredBalance = currentDesiredBalance;
+        final var changes = routingAllocation.changes();
+        final var knownNodeIds = routingAllocation.nodes().stream().map(DiscoveryNode::getId).collect(Collectors.toSet());
+
+        // we assume that all ongoing recoveries will complete
+        for (final var routingNode : routingNodes) {
+            for (final var shardRouting : routingNode) {
+                if (shardRouting.initializing()) {
+                    routingNodes.startShard(logger, shardRouting, changes);
+                    // TODO adjust disk usage info to reflect the assumed shard movement
+                }
+            }
+        }
+
+        // we are not responsible for allocating unassigned primaries of existing shards, and we're only responsible for allocating
+        // unassigned replicas if the ReplicaShardAllocator gives up, so we must respect these ignored shards
+        final RoutingNodes.UnassignedShards unassigned = routingNodes.unassigned();
+        for (final var shardRouting : unassigned) {
+            if (ignoredShards.contains(shardRouting)) {
+                unassigned.ignoreShard(shardRouting, UnassignedInfo.AllocationStatus.NO_ATTEMPT, changes);
+            }
+        }
+
+        // we can assume that all possible shards will be allocated/relocated to one of their desired locations
+        final var unassignedShardsToInitialize = new HashMap<ShardRouting, LinkedList<String>>();
+        for (final var shardAndAssignments : routingNodes.getAssignedShards().entrySet()) {
+            final var shardId = shardAndAssignments.getKey();
+            final List<ShardRouting> shardRoutings = shardAndAssignments.getValue();
+
+            final var shardsToAssign = new ArrayList<ShardRouting>();
+            // treesets so that we are consistent about the order of future relocations
+            final var shardsToRelocate = new TreeSet<>(Comparator.comparing(ShardRouting::currentNodeId));
+            final var targetNodes = new TreeSet<>(desiredBalance.getDesiredNodeIds(shardId));
+            targetNodes.retainAll(knownNodeIds);
+
+            for (ShardRouting shardRouting : shardRoutings) {
+                if (shardRouting.started()) {
+                    if (targetNodes.remove(shardRouting.currentNodeId()) == false) {
+                        shardsToRelocate.add(shardRouting);
+                    }
+                } else {
+                    assert shardRouting.unassigned() : shardRouting;
```

Comment:
Will we ever get here? Looks like we only look at assigned shard routings to begin with/